numpy/_core/numeric.py

```diff
@@ -2385,6 +2385,16 @@ def _array_equal_dispatcher(a1, a2, equal_nan=None):
     return (a1, a2)
 
 
+_no_nan_types = None
+
+
+def _dtype_cannot_hold_nan(dtype):
+    global _no_nan_types
+    if _no_nan_types is None:
+        _no_nan_types = {np.bool_, np.int8, np.int16, np.int32, np.int64}
```

Comment:
Could just use `import .numerictypes as nt`.  The problem is that things are wonky while we are setting things up.
I might slightly prefer (but verbose) to write:
`_no_nan_dtypes = {np.dtypes.BoolDType, np.dtypes.Int8DType, ...}`
and then compare `type(dtype)` rather than `dtype.type`.  We could put the `_no_nan_dtypes` into `np.dtypes` also.

I do *not* want to consider this public, though.  Because the right API should be different (maybe a `np.dtypes.NoNaNDType.register(MyDType)`,

Reply:
Yeah, I agree that the API should not be public, at least for now. And, `numerictypes` seem great !